src/libcore/ptr.rs

```diff
@@ -2556,47 +2549,93 @@ impl<T: ?Sized> NonNull<T> {
     /// The resulting lifetime is bound to self so this behaves "as if"
     /// it were actually an instance of T that is getting borrowed. If a longer
     /// (unbound) lifetime is needed, use `&mut *my_ptr.as_ptr()`.
-    #[stable(feature = "nonnull", since = "1.24.0")]
+    #[stable(feature = "nonnull", since = "1.25.0")]
     pub unsafe fn as_mut(&mut self) -> &mut T {
         &mut *self.as_ptr()
     }
+
+    /// Cast to a pointer of another type
+    #[unstable(feature = "nonnull_cast", issue = /* FIXME */ "0")]
+    pub fn cast<U>(self) -> NonNull<U> where T: Sized {
```

Comment:
What's wrong with `T: ?Sized` here? It makes sense to me to cast the data pointer of `NonNull<[u8]>` to `NonNull<[u8; 64]>` after performing a length check.

Reply:
It's asymmetric because there's no trait that models `as` precisely.

Reply:
Oh you mean the called doing a length check themselves? I added `where T: Sized` because I assumed `as` would not compile without it, but it looks like it does. Updated.